docs(java): standardizing fury java spec (#1240)
apache/incubator-fury: docs/protocols/java_object_graph_spec.md

Format:

```
| length(unsigned varint) | map header | key value pairs data |
```

#### Map header

- For `HashMap/LinkedHashMap`, this will be empty.
- For `TreeMap`, this will be `Comparator`
- For other `Map`, this may be extra object field info.

#### Map Key-Value data

Map iteration is too expensive, Fury can't compute the header like for collection before since it introduce
[considerable overhead](https://github.com/alipay/fury/issues/925).
Users can use `MapFieldInfo` annotation to provide header in advance. Otherwise Fury will use first key-value pair to
predict header optimistically, and update the chunk header if predict failed at some pair.

Fury will serialize map chunk by chunk, every chunk
has 127 pairs at most.

```
+----------------+----------------+~~~~~~~~~~~~~~~~~+
| chunk size: N  |    KV header   |   N*2 objects   |
+----------------+----------------+~~~~~~~~~~~~~~~~~+
```

KV header:

- If track key ref, use first bit `0b1` of header to flag it.
- If key has null, use second bit `0b10` of header to flag it. If ref tracking is enabled for this
  key type, this flag is invalid.
- If map key type is not declared type, use 3rd bit `0b100` of header to flag it.

> - If map key type different, use 4rd bit `0b1000` of header to flag it.

Reviewer: how to understand this sentence? does it means the key is different from all other key?

Author: Every key has a type, for example, if we have a map of type `Map<Foo, String> map`, and we have a subclass of type `class Bar extends Bar`, and map has value `Map.of(new Foo(), "v1", new Bar(), "v2")`. In such cases, map key will have type `Foo` and `Bar`, Fury will need to write key type for every key instead of write it only once.

Reviewer: thanks, now I know that. so actually, the `map key type` should only compare to it's previous sibling key type.

and in some scene, we can use this flag to determine that should we create a new serializer or not.

Author: Yep, and if key type of map is different from previous key, this chunk will be finished, and next chunk may choose to assump every key type of map is different if same key types are not too much

- **Fury Java Binary format**
- **Framework to convert object to/from Fury Java Binary format**

The serialization format is a dynamic binary format. The dynamics and reference/polymorphism support make Fury flexible,
much more easy to use, but
also introduce more complexities compared to static serialization frameworks. So the format will be more complex.

Here is the overall format:

```
| fury header | object ref meta | object class meta | object value data |
```

The data are serialized using little endian byte order overall. If bytes swap is costly for some object,
Fury will write the byte order for that object into the data instead of converting it to little endian.

## Fury header

Fury header consists starts one byte:

```
|     4 bits    | 1 bit | 1 bit | 1 bit  | 1 bit |          optional 4 bytes          |
+---------------+-------+-------+--------+-------+------------------------------------+
| reserved bits |  oob  | xlang | endian | null  | unsigned int for meta start offset |
```

- null flag: 1 when object is null, 0 otherwise. If an object is null, other bits won't be set.

> - endian flag: 1 when the system uses little endian, 0 otherwise.

Reviewer 2: ```suggestion
- endian flag: 1 when data is encoded by little endian, 0 otherwise.
```

Author: This flag is the system endian flag instead of the data. Most data are encoded in little endian. Some arrays such as `long[]` array may use native endian under the hood, we didn't change it to little endian. So if the peer has same endian, we can save the endian conversion cost

Reviewer 2: Ahh I see. Thank you.

Reviewer 2: > Ahh I see. Thank you.

Previously I think that actually there are other endian types rather than big and little, so we'd better describe explicitly.

- xlang flag: 1 when serialization uses xlang format, 0 when serialization uses Fury java format.
- oob flag: 1 when passed `BufferCallback` is not null, 0 otherwise.

If meta share mode is enabled, an uncompressed unsigned int is appended to indicate the start offset of metadata.

## Reference Meta

Reference tracking handles whether the object is null, and whether to track reference for the object by writing
corresponding flags and maintaining internal state.

Reference flags:

| Flag                | Byte Value | Description                                                                                                                              |
|---------------------|------------|------------------------------------------------------------------------------------------------------------------------------------------|

> | NULL FLAG           | `-3`       | This flag indicates the object is a not-null value. We don't use another byte to indicate REF, so that we can save one byte.             |

Reviewer 2: is a null value?

Author: Good catch, updated

> | REF FLAG            | `-2`       | This flag indicates the object is written before, and fury will write a ref id with unsigned varint format instead of serialize it again |

Reviewer 2: ```suggestion
| REF FLAG            | `-2`       | This flag indicates the object is already serialized previously, and fury will write a ref id with unsigned varint format instead of serialize it again |
```

| NOT_NULL VALUE FLAG | `-1`       | This flag indicates the object is a non-null value and fury doesn't track ref for this type of object.                                   |

> | REF VALUE FLAG      | `0`        | This flag indicates the object is a referencable and first write.                                                                        |

Reviewer 2: ```suggestion
| REF VALUE FLAG      | `0`        | This flag indicates the object is referencable and the first time to serialize.                                                                        |
```

When reference tracking is disabled globally or only for some type, or for some type under some context such as some
field of a class, only `NULL FLAG` and ` NOT_NULL VALUE FLAG` will be used.

## Class Meta

> Fury support register class by an optional id, the registration can be used to security check, and identify the class.
> If the class is registered, the class will have a user-provided or an auto-growing unsigned int `class_id`.

Reviewer 2: ```suggestion
Fury supports to register class by an optional id, the registration can be used for security check and class identification.
If a class is registered, it will have a user-provided or an auto-growing unsigned int i.e. `class_id`.
```

      class subsequently. This can reduce array class name cost if component class is or will be serialized.
    - Class will be written as two enumerated fury unsigned by default: `package name` and `class name`. If meta share
      mode is
      enabled,
      class will be written as an unsigned varint which points to index in `MetaContext`.

### Schema evolution

If schema evolution mode is enabled globally or enabled for current class, class meta will be written as follows:

- If meta share mode is not enabled, class meta will be written as schema consistent mode. Additionally, field meta such
  as field type
  and name will be written with the field value using a key-value like layout.
- If meta share mode is enabled, class meta will be written as a meta-share encoded binary if class hasn't been written
  before, otherwise an unsigned varint id which references to previous written class meta will be written.

## Meta share

> This mode will forbid streaming writing since it needs to look back for update the start offset after the whole object
> graph
> writing and meta collecting is finished. Only in this way we can ensure deserialization failure doesn't lost shared
> meta.

> > Meta streamline will be supported in the future for enclosed meta sharing without cross multiple serializations.

Reviewer 2: what's "cross multiple serializations"?

Author: How about changing to:
```
Meta streamline will be supported in the future for enclosed meta sharing without cross multiple serializations of different objects.
```

Reviewer 2: Good to me 👍

inlined
in shared meta header.

Header is written using little endian order, Fury can read this flag first to determine how to deserialize the data.

#### Write by data

If string hasn't been written before, the data will be written as follows:

```
| unsigned varint: string binary size + 1 bit: not written before | 56 bits: unique hash | 3 bits encoding flags + string binary |
```

If string binary size is less than `16` bytes, the hash will be omitted to save spaces. Unique hash can be omitted too
if caller pass a flag to disable it. In such cases, the format will be:

```
| unsigned varint: string binary size + 1 bit: not written before  | 3 bits encoding flags + string binary |
```

#### Write by ref

If string has been written before, the data will be written as follows:

```
| unsigned varint: written string id + 1 bit: written before |
```

## Value Format

### Basic types

#### Bool

- size: 1 byte
- format: 0 for `false`, 1 for `true`

#### Byte

- size: 1 byte
- format: write as pure byte.

#### Short

- size: 2 byte
- byte order: little endian order

#### Char

- size: 2 byte
- byte order: little endian order

#### Unsigned int

- size: 1~5 byte

> - Format: The first bit in every byte indicates whether to have the next byte. If first bit is set

Reviewer 2: ```suggestion
- Format: The most significant bit (MSB) in every byte indicates whether to have the next byte. If first bit is set
```

Reviewer 2: `first` is unclear.

- If the key types of map are different, use the 3rd bit `0b100` of the header to flag it.
- If the actual key type of map is not the declared key type, use the 4rd bit `0b1000` of the header to flag it.
- If track value ref, use the 5th bit `0b10000` of the header to flag it.
- If the value has null, use the 6th bit `0b100000` of the header to flag it. If ref tracking is enabled for this
  value type, this flag is invalid.
- If the value types of map are different, use the 7rd bit `0b1000000` header to flag it.
- If the value type of map is not the declared value type, use the 8rd bit `0b10000000` of the header to flag it.

If streaming write is enabled, which means Fury can't update written `chunk size`. In such cases, map key-value data
format will be:

```
|    1 byte      | variable bytes  |
+----------------+-----------------+
|    KV header   |   N*2 objects   |
```

`KV header` will be a header marked by `MapFieldInfo` in java. For languages such as golang, this can be computed in
advance for non-interface types in most times.

### Enum

Enums are serialized as an unsigned var int. If the order of enum values change, the deserialized enum value may not be

> the value users expect. In such cases, users must register enum serializer by make it write enum value as a enumerated

Reviewer 2: ```suggestion
the value users expect. In such cases, users must register enum serializer by make it write enum value as an enumerated
```

string with unique hash disabled.

### Object

Object means object of `pojo/struct/bean/record` type.
Object will be serialized by writing its fields data in fury order.

Depending on schema compatibility, objects will have different formats.

#### Field order

Field will be ordered as following, every group of fields will have its own order:

- primitive fields: larger size type first, smaller later, variable size type last.

> - boxed primitive fields: same sort as primitive fields

Reviewer 2: ```suggestion
- boxed primitive fields: same order as primitive fields
```

- final fields: same type together, then sorted by field name lexicographically.

> - collection fields: same sort as final fields
> - map fields: same sort as final fields
> - other fields: same sort as final fields

Reviewer 2: ```suggestion
- collection fields: same order as final fields
- map fields: same order as final fields
- other fields: same order as final fields
```

## Spec overview

Fury Java Serialization is an automatic object serialization framework that supports reference and polymorphism. Fury
will
convert an object from/to fury java serialization binary format. Fury has two core concepts for java serialization:

- **Fury Java Binary format**
- Framework to convert object to/from Fury Java Binary format

The serialization format is a dynamic binary format. The dynamics and reference/polymorphism support make Fury flexible,
much more easy to use, but
also introduce more complexities compared to static serialization frameworks. So the format will be more complex.

Here is the overall format:

```
| fury header | object ref meta | object class meta | object value data |
```

> The data are serialized using little endian order overall. If bytes swap is costly, the byte order will be encoded as a

Reviewer: I think the statement should like this, should be more clear:

```diff
- the byte order will be encoded as a flag in data.
+ we can define the endian flag in Fury header
```

Author: How about this:
```
The data are serialized using little endian byte order overall. If bytes swap is costly for some object, 
Fury will write the byte order for that object into the data instead of converting it to little endian.
```

> 

Reviewer 3: The REF FLAG description should indicate whether the 'unsigned ref id' is an unsigned variable integer or an fixed-length integer.

Author: updated

> - If class is registered, it will be written as a little-endian unsigned int: `class_id << 1` using fury unsigned int

Reviewer 4: The concept of 'class_id' was not mentioned or defined elsewhere

Author: Added a description about class id:

Fury support register class by an optional id, the registration can be used to security check, and identify the class.
If the class is registered, the class will have a users provided or a auto-growing id unsigned int `class_id`.

If schema evolution mode is enabled globally or enabled for current class, class meta will be written as follows:

- If meta share mode is not enabled, class meta will be written as scheme consistent mode. Additionally, field meta such
  as field type
  and name will be written when the object value is being serialized using a key-value like layout.
- If meta share mode is enabled, class meta will be written as a meta-share encoded binary if class hasn't been written
  before, otherwise unsigned it will be written as an unsigned varint which references to previous written class meta.

## Meta share

> This mode will forbid streaming writing since it needs to look back for update the offset after the whole object graph
> writing and mete collecting is finished.

> > We have plan to streamline meta writing but haven't started yet.

Author: updated

### Schema consistent

Class will be encoded as an enumerated string by full class name.

### Schema evolution

Class meta format:

```
| meta header: hash + num classes | current class meta | parent class meta | ... |
```

#### Meta header

Meta header is a 64 bits number value encoded in little endian order.

> - Lowest 4 digits `0b0000~0b1110` are used to record num classes. `0b1111` is preserved to indicate that Fury need to

Reviewer 2: max number of classes is 14?

Author: If number of classes  exceeds 14, `0b1111` will be used and another byte will be write next to indicate num classes

Reviewer 2: Ahh I see.

> - If meta share mode is enabled, class will be written as a unsigned int.

Reviewer 4: The sentences feel a little confusing. Perhaps we can change it to:
```
If meta share mode is not enabled, class meta will be written as scheme consistent mode. Additionally， field meta such as field type
  and name will be written when the object value is being serialized using a key-value like layout.
```

Author: updated

information to avoid the cost of writing it for every elements. Specifically, there are four kinds of information
which will be encoded by elements header, each use one bit:

- If track elements ref, use first bit `0b1` of header to flag it.
- If collection has null, use second bit `0b10` of header to flag it. If ref tracking is enabled for this
  element type, this flag is invalid.

> - If collection elements type is not declare type, use 3rd bit `0b100` of header to flag it.

Reviewer 2: what's "declare type" here?

Author: Java support polymorphism, if we declare a field with type `List<SomeClass>`, the element type may be `ChildClass`

Reviewer 2: I see. Maybe you can change it to declared type or compile-time type.

Author: Updated to `declared type`

> - If collection elements type different, use 4rd bit `0b1000` of header to flag it.

Reviewer 2: ```suggestion
- If collection element types are different, use 4rd bit `0b1000` of header to flag it.
```
